cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go

```diff
@@ -27,17 +27,140 @@ import (
 	"math/rand"
 	"net/http"
 	"os"
+	"strings"
 	"text/template"
 	"time"
 
 	"gopkg.in/gcfg.v1"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
 	"k8s.io/autoscaler/cluster-autoscaler/config"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/gpu"
 	klog "k8s.io/klog/v2"
 	schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
 )
 
-type packetManagerRest struct {
+type instanceType struct {
+	InstanceName string
+	CPU          int64
+	MemoryMb     int64
+	GPU          int64
+}
+
+var (
+	maxHttpRetries int           = 3
+	httpRetryDelay time.Duration = 100 * time.Millisecond
+)
+
+// InstanceTypes is a map of packet resources
+var InstanceTypes = map[string]*instanceType{
```

Comment:
I would be very hesitant to hard-code all of these in. The Packet API should provide a way to retrieve a list of all known device types (and if not, I will make sure it gets in).

Reply:
Sure, here it is in [go sdk](https://godoc.org/github.com/packethost/packngo#Plan) and in [rest API](https://www.packet.com/developers/api/plans/)

Reply:
Indeed the Packet API could provide a way to do that, but this data is requested a lot. Hard coded values regarding cloud instances is a common practice with other providers as well (https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/ec2_instance_types.go, https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/alicloud/alicloud_instance_types.go, https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/azure/azure_instance_types.go). Another problem of the Packet API is that it doesn't provide core counts yet, which are needed for this case.

Reply:
> this data is requested a lot

Then we should cache it, even for hours or a day. We don't want to have to cut a new autoscaler release every time packet adds a new node type, or retires one, or bumps the memory on an offering, or any little change.

> Another problem of the Packet API is that it doesn't provide core counts yet, which are needed for this case

We do not? No, we do not, nor do we do the frequency. That is a good reason not to do it. I will open an issue on this for Packet.